std/file.d

```diff
@@ -465,30 +467,98 @@ version (linux) @safe unittest
     assert(read(deleteme) == "abcd");
 }
 
-/********************************************
-Read and validates (using $(REF validate, std,utf)) a text file. $(D S)
-can be a type of array of characters of any width and constancy. No
-width conversion is performed; if the width of the characters in file
-$(D name) is different from the width of elements of $(D S),
-validation will fail.
-
-Params:
-    name = string or range of characters representing the file _name
+/++
+    Read and validates (using $(REF validate, std, utf)) a text file. S can
+    be a type of array of characters of any width and constancy. No width
+    conversion is performed; if the width of the characters in file name
+    is different from the width of elements of S, validation will fail.
+
+    Before validating the characters, the file is first checked for a
+    $(HTTP https://en.wikipedia.org/wiki/Byte_order_mark, BOM). If it is present
+    and matches the encoding of the requested string type, then it is stripped
+    from the output, but if it doesn't match, then a
+    $(REF BOMException, std, encoding) is thrown.
+    $(REF BOMException.bomSeq, std, encoding) will then indicate what the actual
+    encoding was, and the program can then request the string type that matches
+    that encoding.
 
-Returns: Array of characters read.
+    Params:
+        name = string or range of characters representing the file _name
 
-Throws: $(D FileException) on file error, $(D UTFException) on UTF
-decoding error.
- */
+    Returns: Array of characters read.
 
-S readText(S = string, R)(R name)
-if (isSomeString!S &&
-    (isInputRange!R && !isInfinite!R && isSomeChar!(ElementEncodingType!R) || isSomeString!R) &&
-    !isConvertibleToString!R)
+    Throws: $(LREF FileException) if there is an error reading the file,
+            $(REF BOMException, std, encoding) if there is a BOM at the front
+            of the input, and it does not match the requested encoding, and
+            $(REF UTFException, std, utf) if the text contains invalid Unicode.
++/
+S readText(S = string, R)(auto ref R name)
+if (isSomeString!S && (isInputRange!R && !isInfinite!R && isSomeChar!(ElementType!R) || is(StringTypeOf!R)))
 {
-    import std.utf : validate;
+    import std.algorithm.searching : startsWith;
+    import std.encoding : BOM, BOMException;
+    import std.exception : enforce;
+    import std.format : format;
+    import std.utf : UTFException, validate;
+
+    static if(is(StringTypeOf!R))
+        StringTypeOf!R filename = name;
+    else
+        auto filename = name;
+
+    immutable bomSeq = readBOM(filename);
+    immutable bom = bomSeq.schema;
+
+    static if(is(Unqual!(ElementEncodingType!S) == char))
+    {
+        with(BOM) switch(bom)
+        {
+            case none:
+            case utf8: break;
+            default: throw new BOMException(format!"UTF-8 requested, but BOM is %s"(bom), bomSeq);
```

Comment:
This doesn't look right; e.g. it will do the wrong thing if the file starts with the UTF-7 BOM (`+/v8`). This is still valid UTF-8. Instead we probably want to detect cases where the file is not valid in the specified encoding, and throw a better error message than "invalid UTF". IIRC the UTF8/16/32 BOMs are designed to be invalid in all of those encodings except their own, so we can take advantage of that.

Reply:
Are you sure that UTF-7 is valid UTF-8? I didn't think that it was, and `char` is supposed to be UTF-8 by definition.

Reply:
> Are you sure that UTF-7 is valid UTF-8? 

Yes, I'm pretty sure; but that's beside the point. Just the UTF-7 BOM is valid UTF-8, and that's a problem by itself.

> I didn't think that it was, and char is supposed to be UTF-8 by definition.

I don't see how that's related. The problem is that, after this change, you can no longer use `readText` to read text files (which are otherwise valid UTF-8) that start with `+/v8`, which creates an opportunity for a denial of service.

It's not just UTF-7 - look at the `BOM` enum, there's lots of other stuff in there that we don't care about (like EBCDIC).

Reply:
My point is that because `readText` returns a string, and by definition, strings in D are supposed to be UTF-8, UTF-16, or UTF-32 depending on the character type, any other encoding is invalid and should be processed using `read` and turned into `ubyte`s are `uint`s or whatever from there. As it stands, such encodings are likely to result in a `UTFException` being thrown, but if the BOM clearly indicates that the file is not UTF-8, UTF-16, or UTF-32, then there should be no need to call `validate`.

Reply:
Sorry, I think we may be talking past each other.

Here is a D program:

```d
import std.file;

void main()
{
	foreach (string s; ["+/v6", "+/v7", "+/v8"])
	{
		write("test.txt", s);
		readText("test.txt");
	}
}
```

The program is not doing anything invalid. The strings it reads and writes are all perfectly valid UTF-8 (just plain ol' 7-bit ASCII, in fact, which is a subset of UTF-8).

Before this pull request, the program works fine.

After this pull request, the program stops working with:
```
std.encoding.BOMException@.../std/file.d(518): UTF-8 requested, but BOM is utf7
```

This is not a good thing :)

Reply:
Hmm. Well, I guess that we can not throw on UTF-7 then and just let it try and validate it as UTF-8, but really, `readText` is intended for reading UTF-8, UTF-16, and UTF-32 only given that it returns a string.